File: python/08_019.py
from collections import defaultdict, deque

class RateLimiter:
  def __init__(self, message_limit, time_window):
     self.message_limit = message_limit
     self.time_window = time_window
     self.user_messages = defaultdict(deque)
  
  def can_send(self, user_id, timestamp):
     message_queue = self.user_messages[user_id]
     while (message_queue and 
           message_queue[0] <= timestamp - self.time_window):
        message_queue.popleft()
     
     if len(message_queue) < self.message_limit:
        message_queue.append(timestamp)
        return True
     
     return False

File: cpp/08_019.cpp
#include <unordered_map>
#include <deque>
#include <string>

class RateLimiter {
private:
    int messageLimit;
    int timeWindow;
    std::unordered_map<std::string, std::deque<int>> 
        userMessages;
    
public:
    RateLimiter(int messageLimit, int timeWindow) 
        : messageLimit(messageLimit), 
          timeWindow(timeWindow) {}
    
    bool canSend(const std::string& userId, int timestamp) {
        // Get or create the message queue for this user
        auto& messageQueue = userMessages[userId];
        
        // Remove messages outside the current time window
        while (!messageQueue.empty() && 
               messageQueue.front() <= 
                   timestamp - timeWindow) {
            messageQueue.pop_front();
        }
        
        // Check if adding a new message would exceed limit
        if (messageQueue.size() < 
            static_cast<size_t>(messageLimit)) {
            messageQueue.push_back(timestamp);
            return true;
        }
        
        return false;
    }
};